src/expr/src/relation/func.rs

```diff
@@ -709,6 +709,103 @@ where
     })
 }
 
+// The expected input is in the format of [((OriginalRow, InputValue), OrderByExprs...)]
+fn first_value<'a, I>(
+    datums: I,
+    temp_storage: &'a RowArena,
+    order_by: &[ColumnOrder],
+    window_frame: &WindowFrame,
+) -> Datum<'a>
+where
+    I: IntoIterator<Item = Datum<'a>>,
+{
+    // Sort the datums according to the ORDER BY expressions and return the (OriginalRow, InputValue) record
+    let datums = order_aggregate_datums(datums, order_by);
+
+    // Decode the input (OriginalRow, InputValue) into separate datums
+    let datums = datums
+        .into_iter()
+        .map(|d| {
+            let mut iter = d.unwrap_list().iter();
+            let original_row = iter.next().unwrap();
+            let input_value = iter.next().unwrap();
+
+            (input_value, original_row)
+        })
+        .collect_vec();
+
+    let length = datums.len();
+    let mut result: Vec<(Datum, Datum)> = Vec::with_capacity(length);
+    for (idx, (current_datum, original_row)) in datums.iter().enumerate() {
+        let first_value = match &window_frame.start_bound {
+            // Always return the current value
+            WindowFrameBound::CurrentRow => *current_datum,
+            WindowFrameBound::UnboundedPreceding => {
+                if let WindowFrameBound::OffsetPreceding(end_offset) = &window_frame.end_bound {
+                    let end_offset = usize::try_from(*end_offset)
+                        .expect("Window frame offset does not fit in usize");
+
+                    // If the frame ends before the first row, return null
+                    if idx < end_offset {
+                        Datum::Null
+                    } else {
+                        datums[0].0
+                    }
+                } else {
+                    datums[0].0
+                }
+            }
+            WindowFrameBound::OffsetPreceding(offset) => {
+                let start_offset =
+                    usize::try_from(*offset).expect("Window frame offset does not fit in usize");
+                let start_idx = idx.saturating_sub(start_offset);
+                if let WindowFrameBound::OffsetPreceding(end_offset) = &window_frame.end_bound {
+                    let end_offset = usize::try_from(*end_offset)
+                        .expect("Window frame offset does not fit in usize");
+
+                    // If the frame is empty or ends before the first row, return null
+                    if start_offset < end_offset || idx < end_offset {
```

Comment:
I don't see why `idx < end_offset` is needed here. If the first clause is true then we don't care about the second clause. If the first clause is false then the second clause is also always false because `start_idx <= idx` always.

Reply:
The first clause is comparing `start_offset` (the offset given in the `preceding` clause), not the index, so the second clause is not always false. This check is necessary to validate whenever the frame does not include any rows in the partition, i.e. for cases of `ROWS BETWEEN x PRECEDING AND y PRECEDING` where there are no rows in the partition between `x PRECEDING` and `y PRECEDING`.

Here's a SLT test case that fails without this extra check:
```sql
WITH t (f1, f2, f3) AS (VALUES (1, 'a', 1), (2, 'a', 2), (2, 'a', 3), (3, 'a', 4), (4, 'b', 5), (4, 'b', 6), (1, 'c', 7), (2, 'c', 8), (3, 'c', 9), (7, 'd', 10))
SELECT f1, f2, f3, first_value(f1) OVER (PARTITION BY f2 ORDER BY f1, f3 ROWS BETWEEN 10 PRECEDING AND 10 PRECEDING)
FROM t
ORDER BY f2, f3, f1, first_value
```
This returns all nulls on Postgres, but zero nulls without that check.

Here are a couple more examples where you can see the effect of the frame end more clearly:
Postgres:
```sql
WITH t (f1, f2, f3) AS (VALUES (1, 'a', 1), (2, 'a', 2), (2, 'a', 3), (3, 'a', 4), (4, 'b', 5), (4, 'b', 6), (1, 'c', 7), (2, 'c', 8), (3, 'c', 9), (7, 'd', 10))
SELECT f1, f2, f3, first_value(f1) OVER (PARTITION BY f2 ORDER BY f1, f3 ROWS BETWEEN 1 PRECEDING AND 1 PRECEDING)
FROM t
ORDER BY f2, f3, f1, first_value;
 f1 | f2 | f3 | first_value
----+----+----+-------------
  1 | a  |  1 |
  2 | a  |  2 |           1
  2 | a  |  3 |           2
  3 | a  |  4 |           2
  4 | b  |  5 |
  4 | b  |  6 |           4
  1 | c  |  7 |
  2 | c  |  8 |           1
  3 | c  |  9 |           2
  7 | d  | 10 |
```

Mz without that check:
```sql
WITH t (f1, f2, f3) AS (VALUES (1, 'a', 1), (2, 'a', 2), (2, 'a', 3), (3, 'a', 4), (4, 'b', 5), (4, 'b', 6), (1, 'c', 7), (2, 'c', 8), (3, 'c', 9), (7, 'd', 10))
SELECT f1, f2, f3, first_value(f1) OVER (PARTITION BY f2 ORDER BY f1, f3 ROWS BETWEEN 1 PRECEDING AND 1 PRECEDING)
FROM t
ORDER BY f2, f3, f1, first_value;
 f1 | f2 | f3 | first_value
----+----+----+-------------
  1 | a  |  1 |           1
  2 | a  |  2 |           1
  2 | a  |  3 |           2
  3 | a  |  4 |           2
  4 | b  |  5 |           4
  4 | b  |  6 |           4
  1 | c  |  7 |           1
  2 | c  |  8 |           1
  3 | c  |  9 |           2
  7 | d  | 10 |           7
```

Reply:
Ok. The logic took me a while to understand, and apparently I got it wrong anyway.

Reply:
Yeah, unfortunately I didn't find an easier of way of expressing it here. My original idea was to have an auxiliary function that would apply and return the frame, but that ended up being more complicated and harder to work with in cases where frames were empty.